File: Java/Language/Basics/src/Account.java
package banking;

//non-activatable type
public abstract class Account {

	long id;
	protected double balance;

	//pure method (must override)
	public abstract void deposit(double amount);

	public abstract void withdraw(double amount) throws InsufficientFundsException;

	public long getId() { return id; }

	public double getBalance() { return balance; }

	//non-overridable method
	public final void transfer(double amount, Account other) throws InsufficientFundsException {
		if(this == other)
			throw new IllegalTransferException();
		this.withdraw(amount);
		other.deposit(amount);
	}
}

File: CSharp/Basics/InterfaceTest1/Banking/Account.cs
namespace Banking
{
	public abstract class Account
	{
		public long Id { get; internal set; }

		public decimal Balance { get; protected set; }

		public abstract void Deposit(decimal amount);
	
		public abstract void Withdraw(decimal amount);

		public bool Transfer(decimal funds, Account that)
		{
			if(ReferenceEquals(this, that))
				return false;
			this.Withdraw(funds);
			that.Deposit(funds);
			return true;
		}
	}
}

